src/cli/dynamic.rs

```diff
@@ -0,0 +1,333 @@
+        match &self.matches {
+            Some(matches) => target.update_from_arg_matches(matches)?,
+            None => bail!("matches not set"),
+        }
+        Ok(target)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use clap::error::ErrorKind;
+
+    #[derive(Clone, Args)]
+    struct Main {
+        /// Main option
+        #[arg(id = "someopt", long)]
+        opt: Option<String>,
+
+        /// Main flag
+        #[arg(id = "flag", long)]
+        flag: Option<bool>,
+    }
+    impl Default for Main {
+        fn default() -> Self {
+            Main {
+                opt: None,
+                flag: Some(false),
+            }
+        }
+    }
+
+    #[derive(Clone, Args)]
+    struct Mod1 {
+        /// Module 1 someopt
+        #[arg(id = "mod1-someopt", long)]
+        someopt: Option<String>,
+
+        /// Module 1 has a flag true by default
+        #[arg(id = "mod1-flag", long)]
+        flag: Option<bool>,
+    }
+
+    impl Default for Mod1 {
+        fn default() -> Self {
+            Mod1 {
+                someopt: None,
+                flag: Some(true),
+            }
+        }
+    }
+
+    #[derive(Clone, Args)]
+    struct Mod2 {
+        /// Module 2 someopt
+        #[arg(id = "mod2-someopt", long)]
+        someopt: Option<String>,
+
+        /// Module 2 has a flag
+        #[arg(id = "mod2-flag", long)]
+        flag: Option<bool>,
+    }
+
+    impl Default for Mod2 {
+        fn default() -> Self {
+            Mod2 {
+                someopt: None,
+                flag: Some(false),
+            }
+        }
+    }
+
+    #[test]
+    fn dynamic_create() -> Result<()> {
+        assert!(
+            DynamicCommand::new(Main::augment_args(Command::new("mycommand")), "stuff").is_ok()
+        );
+        Ok(())
+    }
+
+    #[test]
+    fn dynamic_register() -> Result<()> {
+        let cmd = DynamicCommand::new(Main::augment_args(Command::new("mycommand")), "stuff");
+        assert!(cmd.is_ok());
+        let mut cmd = cmd?;
+
+        assert!(cmd.register_module::<Mod1>("mod1").is_ok());
+        assert!(cmd.register_module::<Mod1>("mod1").is_err());
+        assert!(cmd.register_module::<Mod2>("mod2").is_ok());
+        assert!(cmd.register_module_noargs("mod3").is_ok());
+        assert!(cmd.register_module_noargs("mod3").is_err());
+        assert!(cmd.modules().contains("mod1"));
+        assert!(cmd.modules().contains("mod2"));
+        assert!(cmd.modules().contains("mod3"));
+        Ok(())
+    }
+
+    #[test]
+    fn dynamic_parse() -> Result<()> {
+        let cmd = DynamicCommand::new(Main::augment_args(Command::new("mycommand")), "stuff");
+        assert!(cmd.is_ok());
+        let mut cmd = cmd?;
+
+        assert!(cmd.register_module::<Mod1>("mod1").is_ok());
+        assert!(cmd.register_module::<Mod2>("mod2").is_ok());
+
+        let err = cmd
+            .command_mut()
+            .try_get_matches_from_mut(vec!["mycommand", "--help"]);
```

Comment:
My idea was to test both _mut and move version of the function. I'm just realizing I should remove _mut from the call below but keep this one to test both ways of matching against a command.
This one has to be kept because the non _mut version consumes (i.e: moves) `self`
https://docs.rs/clap/latest/clap/builder/struct.Command.html#method.try_get_matches_from